security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationManager.scala

```diff
@@ -0,0 +1,107 @@
+    case Some(role) =>
+      authenticationResources.authenticate().optional.flatMap {
+        case Some(user) => provideOrImpersonateUser(user)
+        case None => handleAuthorizationFailedRejection.tmap(_ => AuthenticatedUser.createAnonymousUser(Set(role)))
+      }
+    case None => authenticationResources.authenticate().flatMap(provideOrImpersonateUser)
+  }
+
+  def authenticate(authCredentials: AuthCredentials): Future[Option[AuthenticatedUser]] = authCredentials match {
+    case passedCredentials @ PassedAuthCredentials(_) => authenticationResources.authenticate(passedCredentials)
+    case NoCredentialsProvided                        => authenticateWithAnonymousAccess()
+    case ImpersonatedAuthCredentials(impersonatingUserAuthCredentials, impersonatedUserIdentity) =>
+      authenticateWithImpersonation(impersonatingUserAuthCredentials, impersonatedUserIdentity)
+  }
+
+  def authenticationEndpointInput(): EndpointInput[AuthCredentials] =
+    authenticationResources
+      .authenticationMethod()
+      .and(impersonationHeaderEndpointInput)
+      .map(mappedAuthenticationEndpointInput)
+
+  private def provideOrImpersonateUser(user: AuthenticatedUser): Directive1[AuthenticatedUser] = {
+    optionalHeaderValueByName(impersonateHeaderName).flatMap {
+      case Some(impersonatedUserIdentity) =>
+        authenticationResources.getImpersonatedUserData(impersonatedUserIdentity) match {
+          case Some(impersonatedUserData) =>
+            provide(AuthenticatedUser.createImpersonatedUser(user, impersonatedUserData))
+          case None => reject
+        }
+      case None => provide(user)
+    }
+  }
+
+  private def authenticateWithImpersonation(
+      impersonatingUserAuthCredentials: PassedAuthCredentials,
+      impersonatedUserIdentity: ImpersonatedUserIdentity,
+  ): Future[Option[AuthenticatedUser]] = (for {
+    impersonatingUser <- OptionT(authenticationResources.authenticate(impersonatingUserAuthCredentials))
+    impersonatedUserData <- OptionT(
+      Future { authenticationResources.getImpersonatedUserData(impersonatedUserIdentity.value) }
+    )
+  } yield AuthenticatedUser.createImpersonatedUser(impersonatingUser, impersonatedUserData)).value
+
+  private def handleAuthorizationFailedRejection: Directive0 = handleRejections(
+    RejectionHandler
+      .newBuilder()
+      // If the authorization rejection was caused by anonymous access,
+      // we issue the Unauthorized status code with a challenge instead of the Forbidden
+      .handle { case AuthorizationFailedRejection => authenticationResources.authenticate() { _ => reject } }
+      .result()
+  )
+
+  private def authenticateWithAnonymousAccess(): Future[Option[AuthenticatedUser]] = Future {
+    authenticationResources.getAnonymousRole.map(role => AuthenticatedUser.createAnonymousUser(Set(role)))
+  }
+
+}
+
+object AuthenticationManager {
+  val impersonateHeaderName = "Impersonate-User-Identity"
```

Comment:
I omitted the `X-` prefix on purposie since it's no longer recommended to add it to custom header names https://www.rfc-editor.org/rfc/rfc6648
I added the `Nu` prefix, though.